js/woo.js

```diff
@@ -0,0 +1,1465 @@
+        url += '/' + versionStr + '/';
+        path = this.implodeParams (path, params);
+        params = this.omit (params, this.extractParams (path));
+        params = this.keysort (params);
+        if (privateOrPublic === 'public') {
+            url += privateOrPublic + '/' + path;
+            if (Object.keys (params).length) {
+                url += '?' + this.urlencode (params);
+            }
+        } else {
+            this.checkRequiredCredentials ();
+            url += path;
+            const ts = this.nonce ().toString ();
+            let auth = '';
+            if (method === 'POST' || method === 'DELETE') {
+                auth += this.urlencode (params);
+                body = auth;
+            } else {
+                auth += this.urlencode (params);
+                url += '?' + this.urlencode (params);
+            }
+            auth += '|' + ts;
+            const signature = this.hmac (auth, this.encode (this.secret), 'sha256');
+            headers = {
+                'x-api-key': this.apiKey,
+                'x-api-signature': signature,
+                'x-api-timestamp': ts,
+                'Content-Type': 'application/x-www-form-urlencoded',
+            };
+        }
+        return { 'url': url, 'method': method, 'body': body, 'headers': headers };
+    }
+
+    handleErrors (httpCode, reason, url, method, headers, body, response, requestHeaders, requestBody) {
+        if (!response) {
+            return; // fallback to default error handler
+        }
+        //
+        //     400 Bad Request {"success":false,"code":-1012,"message":"Amount is required for buy market orders when margin disabled."}
+        //
+        const success = this.safeValue (response, 'success');
+        const errorCode = this.safeString (response, 'code');
+        if (!success) {
+            const feedback = this.id + ' ' + this.json (response);
+            this.throwBroadlyMatchedException (this.exceptions['broad'], body, feedback);
+            this.throwExactlyMatchedException (this.exceptions['exact'], errorCode, feedback);
+        }
+    }
+
+    getDefaultNetworkPairForCurrency (code, params) { // TODO: this method can be moved into base
+        // at first, try to find if user or exchange has defined default networks for the specific currency
+        const userChosenNetork = this.safeStringUpper (params, 'network');
+        params = this.omit (params, 'network');
+        const defaultNetworks = this.safeValue (this.options, 'defaultNetworks');
+        if (defaultNetworks !== undefined) {
+            const defaultNetworkCode = this.safeStringUpper (defaultNetworks, code);
+            if (defaultNetworkCode !== undefined) {
+                const networks = this.safeValue (this.options, 'networks');
+                if (networks !== undefined) {
+                    const unifiedNetworkCode = userChosenNetork !== undefined ? this.detectChainSlug (userChosenNetork) : defaultNetworkCode;
```

Comment:
This line is a violation of the following rule: 

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#using-ternary-conditionals